src/glpk/env/alloc.c

```diff
@@ -45,74 +46,141 @@
+
+/**
+ * "Unwrap" the memory block descriptor
+ */
+static inline MBD* ptr_to_mbd(void* ptr)
+{
+    MBD* mbd = (MBD *)((char *)ptr - MBD_SIZE);
+    xassert(mbd->self == mbd);
+    return mbd;
+}
+
+/**
+ * Take the mbd off the environment. Must only ever try to remove an mdb from the environment
+ * it was allocated on
+ */
+static inline void _remove_from_env(MBD* mbd, ENV* env)
+{
+    xassert(mbd->env == env);
+
+    MBD* before = mbd->prev;
+    MBD* after = mbd->next;
+
+    /* unlink ourselves */
+    if(after) {
+        after->prev = before;
+    }
+    if(before) {
+        before->next = after;
+    } else {
+        env->mem_ptr = after;
+    }
+
+    /* update counters */
+    if (get_mem_count() < 1 || get_mem_total() < mbd->size) {
+        xerror("unlinking mbd: memory allocation error\n");
+    }
+    add_mem_count(-1);
+    add_mem_total(-(mbd->size));
+}
+
+/**
+ * Free an mbd from an env. If the mbd's env and the current env mismatch, the block will *not* be deallocated yet,
+ * instead the deallocation will be deferred until the environment is freed
+ */
+static inline void* _free_mbd(MBD* mbd, ENV* env) {
+    if(mbd->env != env) {
+        return NULL;
+    }
+    _remove_from_env(mbd, env);
+    free(mbd);
+    return NULL;
+}
+
+static inline void _check_allocation(ENV* env, size_t size) 
+{
+    if (size > SIZE_T_MAX - MBD_SIZE)
+        xerror("block too large\n");
+    if (size + MBD_SIZE > get_mem_limit() - get_mem_total())
+        xerror("memory allocation limit exceeded\n");
+    if (get_mem_count() == INT_MAX)
+        xerror("too many memory blocks allocated\n");
+}
+
+static inline void _prepend_mbd_to_env(MBD* mbd, ENV* env, size_t size)
+{
+    xassert(mbd != NULL && env != NULL);
+    mbd->self = mbd;
+    mbd->size = size;
+    mbd->env = env;
+    mbd->prev = NULL;
+    mbd->next = env->mem_ptr;
+
+    if(mbd->next == NULL) {
+        env->mem_tail_ptr = mbd;
+    } else {
+        mbd->next->prev = mbd;
+    }
+
+    env->mem_ptr = mbd;
+    add_mem_count(1);
+    set_mem_cpeak();
+
+    add_mem_total(size);
+    set_mem_tpeak();
+}
+
+/**
+ * Calloc an mbd and add it to the env
+ */
+static inline void* _calloc_mbd(ENV* env, size_t size)
+{
+    MBD* mbd = NULL;
+    if(size == 0) {
+        return NULL;
+    }
+    _check_allocation(env, size);
+    xassert((mbd = calloc(1, size + MBD_SIZE)) != NULL);
+    _prepend_mbd_to_env(mbd, env, size);
+    return mbd;
+}
+
+/**
+ * Reallocate an mdb. This can *only* work if the mbd was previously allocated in this env. 
+ * (if size is 0, will free the mbd instead)
+ */
+static inline void* _realloc_mbd(MBD* mbd, ENV* env, size_t size)
+{
+    if(size == 0) {
+        return _free_mbd(mbd, env);
+    }
+    if(mbd->env != env) {
+        xerror("%s: ptr = %p; unable to reallocate from another environment", mbd);
+    }
+    _check_allocation(env, size);
+    _remove_from_env(mbd, env);
+    xassert((mbd = realloc(mbd, size + MBD_SIZE)) != NULL);
```

Comment:
```suggestion
        mbd = realloc(mbd, size + MBD_SIZE);
        xassert(mdb != NULL);
```

Reply:
I think this is easier to read if it is split into two statements. Also, I've seen compiler directives that will #define away assert statements. Not sure if this library supports that, but if it does then the call to realloc() will get discarded.